docs/RFCS/data_structures_for_logical_planning.md

```diff
@@ -0,0 +1,1001 @@
+     .------'    (exists)
+     |               \
+     |              (filter)
+     |             /       \
+ (scan "foo")  (scan "t")   '--.
+     |            |            |
+  (var "x")    (var "a")      (=)
+     \            \          /  |
+      \            '--------'   |
+       '------------------------'
+```
+
+## Early advantages
+
+Supposing we can use a single Go type to encode all the various
+expression representations (this is made possible by the [ADT
+RFC](20170517_algebraic_data_types.md)), the memo can be highly
+compact in memory and keep the number of allocations much lower than
+currently.
+
+### Where the magic happens, part 1: conjunctions
+
+Suppose we also consider AND expressions and (in)equality
+comparisons. For this we need the new m-expression type `(and z)`
+appears. `z` is an *integer set*, which we can conveniently encode as
+a bitmap.
+
+For example:
+
+```sql
+select x from foo where x != 0 and x < 10 and x != 0
+```
+
+Memo:
+
+```
+0: (project 1 3)
+1: (scan "foo" 2)
+2: (var "x")
+3: (and {7,4})
+4: (and {7,5})
+5: (< 2 6)
+6: (literal "10")
+7: (= 2 8)
+8: (literal "0")
+```
+
+Now here's where the magic happens: in the simplify phase, we can
+always apply the rule `(and (and x) y) = (and x (and y)) = (and (union
+x y))`. With the integer sets represented as bitmaps, we just need to
+compute a bitwise OR of the bitmaps. There is no memory traffic. And
+then the two `(and)` representations in the example collapse to:
+
+```
+0: (project 1 3)
+1: (scan "foo" 2)
+2: (var "x")
+3: (and {7,5}) (and {7,4}) -- the two m-expressions are equivalent
+4: (and {7,5})
+5: (< 2 6)
+6: (literal "10")
+7: (= 2 8)
+8: (literal "0")
+```
+
+This optimization is particularly important during predicate
+push-down, as will be detailed later.
+
+### Where the magic happens, part 2: multicolumn rel exprs
+
+Suppose we also consider multicolumn queries.
+
+Then:
+
+- the representation type `(scan <name> r)` must be extended so
+  `r` is an index list (to decide in which order the columns on disk are
+  presented in the resulting placeholders)
+
+- the representation type `(project x r)` is extended accordingly,
+  with `r` an index list instead of a single index as previously. `x`
+  is the source relational expression as previously.
+
+Then:
+
+```sql
+select k, v from kv
+```
+
+Memo:
+
+```
+0: (project 1 [2 3])
+1: (scan "kv" [2 3])
+2: (var "k")
+3: (var "v")
+```
+
+Another:
+
+```sql
+select v, k from (table "kv")
+```
+
+Memo:
+
+```
+0: (project 1 [3 2])
+1: (scan "kv" [2 3])
+2: (var "k")
+3: (var "v")
+```
+
+Another:
+
+```sql
+select k+1,k+1 from kv
+```
+
+Memo:
+
+```
+0: (project 1 [4 4])
+1: (scan "kv" [2 3])
+2: (var "k")
+3: (var "v")
+4: (+ 2 5)
+5: (literal "1")
+```
+
+See how common (sub-)expressions can be shared.
+
+### Where the magic happens, part 3: stacked projections
+
+Example:
+
+```sql
+select k, v from (select v, k from kv)
+```
+
+Memo:
+
+```
+0: (project 1 [3 4])
+1: (project 2 [4 3])
```

Comment:
Ah! Here I understand why you need `project` (by attaching it to the data source, you get to reuse the `var`).